payroll/views.py

```diff
@@ -177,42 +177,64 @@ def get_context_data(self, **kwargs):
+        return value
+
+
 class DownloadView(TemplateView):
     def get(self, request, *args, **kwargs):
         slug = request.GET.get('employer')
         year = request.GET.get('year')
         employer = Employer.objects.get(slug=slug)
-        employer_salaries = employer.get_salaries(year=year)
+        employer_salaries = employer.get_salaries(year=year).select_related(
+            'job',
+            'job__person',
+            'job__position',
+            'job__position__employer',
+            'job__position__employer__parent'
+        )
         
-        buffer = StringIO()
-        header = ['name', 'unit', 'department', 'title', 'tenure', 'salary', 'overtime']
-        dict_writer = csv.DictWriter(f=buffer, fieldnames=header)
-        dict_writer.writeheader()
-
-        for salary in employer_salaries:
-            name_kwargs = {
-                'first_name': salary.job.person.first_name,
-                'last_name': salary.job.person.last_name
-            }
-            name = '{first_name} {last_name}'.format(**name_kwargs)
-
-            start_date = salary.job.start_date.strftime('%m/%d/%Y') if salary.job.start_date else ''  # noqa
-
-            dict_writer.writerow({
-                'name': name,
-                'unit': employer.parent,
-                'department': employer.name,
-                'title': salary.job.position.title,
-                'tenure': start_date,
-                'salary': salary.amount,
-                'overtime': salary.extra_pay
-            })
-
-        buffer.seek(0)
-        response = HttpResponse(buffer, content_type='text/csv')
-        filename = '{employer}-{year}.csv'.format(employer=employer.name, year=year)  # noqa
+        buffer = PsuedoBuffer()
+        # todo: don't worry ab this rn. DictWriter requires the argument so leaving it here for now
+        headers = ['name', 'unit', 'department', 'title', 'tenure', 'salary', 'overtime']
+        dict_writer = csv.DictWriter(f=buffer, fieldnames=headers)
+
+        # todo: move outside the scope of this `get` method?
+        def row_generator():
+            for salary in employer_salaries:
+                name_kwargs = {
+                    'first_name': salary.job.person.first_name,
+                    'last_name': salary.job.person.last_name
+                }
+                name = '{first_name} {last_name}'.format(**name_kwargs)
+
+                start_date = salary.job.start_date.strftime('%m/%d/%Y') if salary.job.start_date else ''  # noqa
+                
+                yield {
+                    'name': name,
+                    'unit': employer.parent,
+                    'department': employer.name,
+                    'title': salary.job.position.title,
+                    'tenure': start_date,
+                    'salary': salary.amount,
+                    'overtime': salary.extra_pay
+                }
+                
+        rows = row_generator()
```

Comment:
there's a few different ways to get the header in there. I think this approach seems pretty good to me.

itertools has a bunch of neat stuff, including the ability to chain two iterables together without evaluating them until they are needed.

```python
import itertools
rows = itertools.chain([dict_writer.writeheader()],
                        (dict_writer.writerow(row) for row in row_generator())

response = StreamingHttpResponse(
            rows,
            content_type='text/csv'
        )       
```

Reply:
@fgregg changed with the two most recent commits. thanks for pairing on that.